synapse/static/client/login/js/login.js

```diff
@@ -46,23 +48,36 @@ var setFeedbackString = function(text) {
 };
 
 var show_login = function() {
-    $("#loading").hide();
-
     var this_page = window.location.origin + window.location.pathname;
     $("#sso_redirect_url").val(this_page);
 
-    if (matrixLogin.serverAcceptsPassword) {
-        $("#password_flow").show();
+    if (matrixLogin.serverAcceptsSso) {
+        if (try_token() || matrixLogin.serverAcceptsPassword) {
```

Comment:
rather than this (which will try a second attempt to submit the token), I think it would make sense to give show_login an `inhibit_redirect` flag, which is true when it is called from `errorFunc` and false when it is called from `onLoad`.